mobilecoind/src/payments.rs

```diff
@@ -324,16 +365,184 @@ impl<T: BlockchainConnection + UserTxConnection + 'static, FPR: FogPubkeyResolve
         };
         log::trace!(logger, "Tombstone block set to {}", tombstone_block);
 
+        // Convert outlays into OutlayV2, using fee token id to set the token id.
+        let outlays: Vec<OutlayV2> = outlays
+            .into_iter()
+            .map(|outlay_v1| OutlayV2 {
+                receiver: outlay_v1.receiver.clone(),
+                amount: Amount::new(outlay_v1.value, token_id),
+            })
+            .collect();
+
         // Build and return the TxProposal object
         let mut rng = rand::thread_rng();
         let tx_proposal = Self::build_tx_proposal(
             &selected_utxos_with_proofs,
             rings,
+            &[],
             block_version,
             token_id,
             fee,
             &sender_monitor_data.account_key,
             change_subaddress,
+            &outlays,
+            tombstone_block,
+            &self.fog_resolver_factory,
+            opt_memo_builder,
+            fee_map,
+            &mut rng,
+            &self.logger,
+        )?;
+        log::trace!(logger, "Tx constructed, hash={}", tx_proposal.tx.tx_hash());
+
+        Ok(tx_proposal)
+    }
+
+    /// Create a TxProposal, possibly with mixed token ids.
+    ///
+    /// # Arguments
+    /// * `sender_monitor_id` - Indicates the the account key needed to spend
+    ///   the txo's.
+    /// * `fee_token_id` - The token id to transact in.
+    /// * `change_subaddress` - Recipient of any change.
+    /// * `inputs` - UTXOs that will be spent by the transaction.
+    /// * `scis` - SCIs to incorporate into the transaction
+    /// * `outlays` - Output amounts and recipients.
+    /// * `last_block_infos` - Last block info responses from the network, for
+    ///   determining fees. This should normally come from polling_network_state
+    /// * `opt_fee` - Transaction fee in picoMOB. If zero, defaults to MIN_FEE.
+    /// * `opt_tombstone` - Tombstone block. If zero, sets to default.
+    /// * `opt_memo_builder` - Optional memo builder to use instead of the
+    ///   default one (EmptyMemoBuilder).
+    pub fn build_mixed_transaction(
```

Comment:
One place we could make an executive decision here is, instead of having `build_transaction` and `build_mixed_transaction` in this file, we could just have `build_mixed_transaction`, and the current callers of `build_transaction` could call this one instead, since it's a superset of the functionality.

Such a change has a larger code footprint on the existing APIs, but since `build_tx_proposal` has also been extended in this way, likely we should just do the same thing here I guess.

I will likely do this in the next commit.